lib/report.js

```diff
@@ -158,7 +160,144 @@ class ConciseReporter extends Reporter {
   }
 }
 
+class TapReporter extends Reporter {
+  constructor(options) {
+    super(options);
+    this.successful = 0;
+    this.failed = 0;
+    this.skipped = 0;
+    this.failedTests = [];
+    this.testNumeration = 1;
+    this.message = '';
+    this.got = [];
+    this.expected = [];
+  }
+
+  error(test, error) {
+    // remove one success because 'error' may only be received after end
+    // though the test may have already failed so we need a way to check
+    // if we should do this, i.e. store map of successful tests
+    // this.successes--;
+    // test.todo ? this.skipped++ : this.failures++;
+    // TODO(lundibundi): add test.metadata.filename here too once available
+    this.listFailure(
+      test,
+      { message: `Failure after end: ${error ? error.message : 'no message'}` },
+      ` Test failed: ${test.caption}`
+    );
+  }
+
+  parseTestResults(test, offset = 0) {
+    if (!test.success) {
+      if (!test.todo) {
+        this.failedTests.push(this.testNumeration);
+      }
+      this.listFailure(test, true);
+      this.got.forEach(el => {
+        this.message += `${pad(6)}- ${el}\n`;
+      });
+      this.message += `${pad(4)}expect: \n`;
+      this.expected.forEach(el => {
+        this.message += `${pad(6)}- ${el}\n`;
+      });
+      this.message += '  ...';
+      this.got = [];
+      this.expected = [];
+      console.log(this.message);
+    } else {
+      if (offset === 0) this.testNumeration++;
+      test.results.forEach((res, index) => {
+        const testId = index + 1;
+        if (res.success) {
```

Comment:
Why are all not successful assertions ignored?

Reply:
Because if not successful assertions are in the test, so it'll be not ok or failed.